File: flask_app/app/controllers/discount_controller.py
from flask import request
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.services.discount_service import DiscountService
from app.utils.response import success_response, error_response
from app.utils.decorators import admin_required

class DiscountController:

    @staticmethod
    @jwt_required()
    def get_all():
        page = request.args.get('page', default=1, type=int)
        per_page = request.args.get('per_page', default=10, type=int)
        
        filters = {
            "discount_type":request.args.get('discount_type'),
            "active":request.args.get("active", type=lambda v: v.lower() == 'true' if v else None),
        }

        filters = {k: v for k, v in filters.items() if v is not None}

        discounts, total = DiscountService.get_all(page, per_page, filters = filters)
        return success_response("Discounts list", {
            "items": [d.to_dict() for d in discounts],
            "total": total,
            "page": page,
            "per_page": per_page,
        })

    @staticmethod
    @jwt_required()
    def get_one(discount_id):
        discount = DiscountService.get_by_id(discount_id)
        if not discount:
            return error_response("Discount not found", 404)
        return success_response("Discounts Detail",discount.to_dict())
    
    @staticmethod
    @jwt_required()
    def get_by_code(code):
        discount = DiscountService.get_by_code(code)
        if not discount:
            return error_response("Discount not found", 404)
        return success_response("Discounts Detail",discount.to_dict())

    @staticmethod
    @admin_required()
    def create():
        data = request.get_json()
        try:
            discount, success = DiscountService.create(data)
            if success:
                return success_response("Discount Created",discount.to_dict(), 201)
            else:
                return error_response(discount, 400)
            
        except Exception as e:
            return error_response(str(e), 400)

    @staticmethod
    @jwt_required()
    def validate_discount():
        data = request.get_json()
        valid, msg = DiscountService.validate_discount(data)
        return success_response(msg, {"valid": valid})

    @staticmethod
    @admin_required()
    def update(discount_id):
        data = request.get_json()
        try:
            discount = DiscountService.update(discount_id, data)
            return success_response("Discount Updated",discount.to_dict())
        except Exception as e:
            return error_response(str(e), 400)

    @staticmethod
    @admin_required()
    def delete(discount_id):
        try:
            DiscountService.soft_delete(discount_id)
            return success_response({"message": "Deleted successfully"})
        except Exception as e:
            return error_response(str(e), 400)

File: java_membership_api/src/main/java/com/example/membership_api/controller/DiscountController.java
package com.example.membership_api.controller;

import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.membership_api.annotations.AdminOnly;
import com.example.membership_api.model.Discount;
import com.example.membership_api.service.DiscountService;
import com.example.membership_api.utils.ResponseUtil;

@RestController
@RequestMapping("/discounts")
public class DiscountController {

    @Autowired
    private DiscountService discountService;

    @GetMapping("")
    public Object getAll(
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int perPage,
            @RequestParam Map<String, Object> filters
    ) {
        try {
            Map<String, Object> result = discountService.getAll(page, perPage, filters);
            return ResponseUtil.success("Discounts fetched successfully", result);
        } catch (Exception e) {
            return ResponseUtil.error("Error fetching discounts: " + e.getMessage());
        }
    }

    @GetMapping("/{id}")
    public Object getById(@PathVariable String id) {
        Discount discount = discountService.getById(id);
        if (discount == null) {
            return ResponseUtil.error("Discount not found", 404);
        }
        return ResponseUtil.success("Discount fetched successfully", discount);
    }

    @GetMapping("/code/{code}")
    public Object getByCode(@PathVariable String code) {
        Discount discount = discountService.getByCode(code);
        if (discount == null) {
            return ResponseUtil.error("Discount not found", 404);
        }
        return ResponseUtil.success("Discount fetched successfully", discount);
    }

    @PostMapping("")
    @AdminOnly
    public Object create(@RequestBody Map<String, Object> data) {
        try {
            Object[] result = discountService.create(data);
            if ((boolean) result[1]) {
                return ResponseUtil.success("Discount created successfully", result[0]);
            } else {
                return ResponseUtil.error((String) result[0]);
            }
        } catch (Exception e) {
            return ResponseUtil.error("Error creating discount: " + e.getMessage());
        }
    }

    @PutMapping("/{id}")
    @AdminOnly
    public Object update(@PathVariable String id, @RequestBody Map<String, Object> data) {
        try {
            Discount updated = discountService.update(id, data);
            return ResponseUtil.success("Discount updated successfully", updated);
        } catch (Exception e) {
            return ResponseUtil.error("Error updating discount: " + e.getMessage());
        }
    }

    @DeleteMapping("/{id}")
    @AdminOnly
    public Object delete(@PathVariable String id) {
        try {
            boolean deleted = discountService.softDelete(id);
            if (deleted) {
                return ResponseUtil.success("Discount deleted successfully");
            } else {
                return ResponseUtil.error("Discount not found", 404);
            }
        } catch (Exception e) {
            return ResponseUtil.error("Error deleting discount: " + e.getMessage());
        }
    }

    @PostMapping("/validate")
    public Object validateDiscount(@RequestBody Map<String, Object> payload) {
        try {
            Object[] result = discountService.validateDiscount(payload);
            boolean isValid = (boolean) result[0];
            String message = (String) result[1];
            if (isValid) {
                return ResponseUtil.success(message);
            } else {
                return ResponseUtil.error(message);
            }
        } catch (Exception e) {
            return ResponseUtil.error("Error validating discount: " + e.getMessage());
        }
    }
}
